### ISSUE TITLE
registerPlugin 

### ISSUE THREAD

**Issue by anaraserikbek** [RGLab/openCyto#114]
Hi, I'm trying to create a plugin function for working with it in the .csv file. 
However, when I try to register my function with the following code: 

```
> gmmclustering <- function(fr, pp_res, channels = NA, ...){
> 
>       Mclust:::Mclust(fr, channels = channels, ...)
>    }
> registerPlugins(fun=gmmgating, methodName=Mclust, character=gating)
```

I receive `Error: could not find function "registerPlugins"` 
I updated openCyto and there is still no change.

**Comment by gfinak** [comment]
Pardon my silly question, but did you load the openCyto library first?
Also it looks like you should be registering `gmclustering` not `gmgating`,
`methodName` should be `"Mclust"` (in quotes, a character vector, not a bare variable name),
Finally, there is no argument `character` to the function, it should by `type="gating"`, `type` is passed on through the `...` to the `.register` function.

**Comment by anaraserikbek** [comment]
hi Grek, thank you. 
yes, i have of-course loaded the openCyto & flowStats libraries first. 
i'm sorry about the registering function name, yes it should be gmmclustering. 
my final function looks like this: 
_```
gmmclustering <- function(fr, pp_res, channels = NA, ...){

      Mclust:::Mclust(fr, channels = channels, ...)
   }
registerPlugins(fun=gmmclustering, methodName="Mclust", type="gating")
```_
 
and I still receive an error: _Error: could not find function "registerPlugins"_

**Comment by gfinak** [comment]
Could you double check the version of open Cyto against what's in
BioConductor development branch?

Greg Finak

**Comment by mikejiang** [comment]
Your `openCyto` was not installed. Try `biocLite('openCyto')` again and save the entire error message as text and attach it here

**Comment by anaraserikbek** [comment]
Okay, so when I deleted and re-installed openCyto through biocLite by `biocLite('openCyto')`, I received the following message: 
```
BioC_mirror: https://bioconductor.org
Using Bioconductor 3.2 (BiocInstaller 1.20.3), R 3.2.5 (2016-04-14).
Installing package(s) ‘openCyto’
trying URL 'https://bioconductor.org/packages/3.2/bioc/bin/windows/contrib/3.2/openCyto_1.8.4.zip'
Content type 'application/zip' length 2041388 bytes (1.9 MB)
downloaded 1.9 MB

package ‘openCyto’ successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\amnis\AppData\Local\Temp\Rtmp0gjAu1\downloaded_packages
```
Then, I did a sessionInfo() & I received the following list of all the packages, where openCyto was not there: 
```
R version 3.2.5 (2016-04-14)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 7 x64 (build 7601) Service Pack 1

locale:
[1] LC_COLLATE=English_United States.1252 
[2] LC_CTYPE=English_United States.1252   
[3] LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] splines   stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] ggplot2_2.1.0             flowWorkspace_3.16.16    
 [3] mclust_5.2                flowWorkspaceData_2.6.1  
 [5] sgeostat_1.0-27           fda_2.4.4                
 [7] Matrix_1.2-6              gridExtra_2.2.1          
 [9] ncdfFlow_2.16.1           BH_1.60.0-2              
[11] RcppArmadillo_0.7.100.3.1 flowViz_1.34.1           
[13] lattice_0.20-33           flowCore_1.36.9          
[15] BiocInstaller_1.20.3     
```
I also tried to then again double check and did `library(openCyto)` and received this: 

```
Error in loadNamespace(j <- i[[1L]], c(lib.loc, .libPaths()), versionCheck = vI[[j]]) : 
  there is no package called ‘GGally’
Error: package or namespace load failed for ‘openCyto’

```
and when I do a traceback after this, I receive: 

```
2: stop(gettextf("package or namespace load failed for %s", sQuote(package)), 
       call. = FALSE, domain = NA)
1: library(openCyto)
```


However, the bizarre thing is that openCyto folder with all the usual information is in the folder of C/R/libraries. 
```
I know my R is an older version, but with the PC I'm currently using it's the highest i can get. 


Best,
Anara

**Comment by mikejiang** [comment]
These are silly to ask, but
If it tells you `GGally` is missing, then what about to install it?
And are you sure your sever is connected to internet?

**Comment by SamGG** [comment]
Hi,
Got same error on my Windows with R 3.2.3. When reinstalling GGally, R tends to install from source because source package is more recent. I am unsure about which package require GGally, but to make the story short, try `install.packages("GGally", type = "win.binary")`. It works for me.
Concerning your server, I think Mike is right: either it's not connected, either you don't have enough privileges.
Best.

**Comment by mikejiang** [comment]
![image](https://cloud.githubusercontent.com/assets/1385649/16435509/5169aa84-3d4b-11e6-86ed-2e033ff55553.png)
![image](https://cloud.githubusercontent.com/assets/1385649/16435613/36282cd6-3d4c-11e6-8b11-ceb31cbd6457.png)
The two graphs show that `openCyto` does not directly use `GGally`. Its is `flowStats->mvoutlier-->robCompositions-->GGally` that causes this dependencies. I will see if I can eliminate this requirement by inline some code from `mvoutlier`.

**Comment by SamGG** [comment]
@mikejiang IMPRESSIVE!
I think that this error shows the limits of "Don't reinvent the wheel".

**Comment by anaraserikbek** [comment]
@mikejiang  yes of-course I've tried to install GGally both from Bioconductor (through biocLite and I'v e receive the following & the package is not in the library:
```
Error in download.file(url, destfile, method, mode = "wb", ...) : 
  cannot open destfile 'C:\Users\amnis\AppData\Local\Temp\Rtmp6jufEj/downloaded_packages/GGally_1.1.0.tar.gz', reason 'Permission denied'
Warning in download.packages(pkgs, destdir = tmpd, available = available,  :
  download of package ‘GGally’ failed
```
and when I do `traceback `on this: 
```
2: stop(txt, domain = NA)
1: library(GGally)
```

-- However, since it's a CRAN package, I have also tried to install it though CRAN, by `install.packages()` and I still could not see it in the library. 

-- I also have tried to install GGally directly download the .zip file and putting the unzipped package file physically directly in the library, however even if R saw the package it still did not read, giving this error:  `Error in library("GGally") : ‘GGally’ is not a valid installed package`



P.S. what I don't really understand is that openCyto was functioning before & i had never a problem with the GGally and other packages being missing for openCyto while installing. the openCyto package file is in the library, but as I said it isn't opened in R. 


Also, i've checked, but other packages seem to be functioning, so it shouldn't be a problem of the library. 

I wonder if this could happen if some packages are updated & the older version of GGally is not functional for current openCyto? 

P.S. currently I'll test another version of R (I have 3.2.5 right now) & will test 3.3.1. 
Also, i will test if older versions of openCyto could be working. Where can I get a very old (2014-ish) version of openCyto?


thank you  for your help @SamGG and Mike
